cpp/cycles_module/algorithm/cycles.cpp

```diff
@@ -0,0 +1,229 @@
+#include <algorithm>
+#include <cassert>
+#include <set>
+#include <unordered_set>
+
+#include <mg_exceptions.hpp>
+#include <mg_graph.hpp>
+
+#include "cycles.hpp"
+
+namespace cycles_util {
+
+NodeState::NodeState(size_t number_of_nodes) {
```

Comment:
Just to be consistent, since we are using everything from std, we could also use std::size_t
```suggestion
NodeState::NodeState(std::size_t number_of_nodes) {
```